### PR TITLE
Prefer custom SMTP config over Sendgrid

### PR DESCRIPTION
**Pull request by maestromac**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
We will now give the user's custom SMTP setting a priority over Sengrid settings.

I believe only checking for `address` is enough? I'm not sure what's the bare minimum needed to configure SMTP but if there's a mailing server with no authentication whatsoever, address is enough?

## Related Tickets & Documents
a part of https://github.com/forem/forem/issues/16180

## QA Instructions, Screenshots, Recordings
It's a straightforward logic swap with specs to back it up.

## Added tests?
- [x] no, because they aren't needed

## Added to documentation?
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?
n/a

### CODE REVIEW

File: app/models/settings/smtp.rb
Comment (Ridhwana) [inline app/models/settings/smtp.rb:17]: In response to your question in the description, I also don't know whats the bare minimum but should we align this with our ForemInstance.smtp_enabled? method where we check if the the username and password is present?

```

  def self.smtp_enabled?
    (Settings::SMTP.user_name.present? && Settings::SMTP.password.present?) || ENV["SENDGRID_API_KEY"].present?
  end
```

Comment (maestromac) [reply app/models/settings/smtp.rb:17]: that's a good idea. let me do that.

Comment (maestromac) [reply app/models/settings/smtp.rb]: updated!

File: app/models/settings/smtp.rb
Comment (citizen428) [inline app/models/settings/smtp.rb:31]: Suggestion: this may be a bit clearer. It also uses core Ruby and no Rails extensions (`index_with` was added to Rails before the block for `to_h` was added to Ruby):

```suggestion
        keys.to_h { |k| [k.to_sym, public_send(k)] }
```

Comment (citizen428) [reply app/models/settings/smtp.rb:31]: Also, this gave me the idea to maybe expose this for all settings models as part of `Settings::Base`.

**Review by citizen428 — APPROVED** [review]

File: app/models/settings/smtp.rb
Comment (citizen428) [reply app/models/settings/smtp.rb:31]: PR is up, we can change this after merging to simply call `to_h`.

**Review by Ridhwana — APPROVED** [review]
Approving because the code looks good. However, theres a failing test.

**Comment by Ridhwana** [comment]
@maestromac I just thought of a scenario now: 

What if someone is on a Forem Cloud account, and they have sendgrid setup. They choose to override their sendgrid account with their own SMTP settings. The change that you've put into place will ensure that they can do this. 

Now what happens if they want to revert those changes, i.e. move from SMTP Settings back to Sendgrid, can they do this by just clearing out the fields? I _think_ theres validation on the fields so that may not work.

**Comment by maestromac** [comment]
@Ridhwana I just tested locally and it kinda works. You can erase all the fields but the authentication method, which as you mentioned is relying on validation to include something among an array. Should we address that in a different PR?

**Comment by Ridhwana** [comment]
> @Ridhwana I just tested locally and it kinda works. You can erase all the fields but the authentication method, which as you mentioned is relying on validation to include something among an array. Should we address that in a different PR?

yeah lets do that, it should definitely not hold this PR back or anything.

File: app/models/settings/smtp.rb
Comment (citizen428) [reply app/models/settings/smtp.rb:31]: Just FYI, the other PR has been merged now in case you want to update but we can always do that later too.

Comment (maestromac) [reply app/models/settings/smtp.rb]: Updated!

File: app/models/settings/smtp.rb
Comment (maestromac) [inline app/models/settings/smtp.rb:32]: This feels a bit overdone but I like the explicit description of the method.

Comment (citizen428) [reply app/models/settings/smtp.rb:32]: It is more maintable. If we ever add new settings on the model that don't need to be included we can simply update this to `to_h.except(...)` and not change any call sites.

**Pull request merged**
